[refactor] 강아지 생성/ 강아지 주인전화번호로 조회/ 강아지 진료기록 추가 함수에 MongoDB적용

qkrdmstlr3 opened the pull request:
## 변경사항

- api호출을 dog에서 dogs로 경로 변경
  - REST API에서 자원은 복수여야하기 때문이다
- 강아지 생성 함수 추가
  - 이름/소유자이름/소유자번호가 중첩되지 않도록 필터링을 해줌
  - 강아지가 기존에 있다면 409에러 반환
- 강아지 주인번호로 조회 함수 변경
  - mongodb알고리즘 추가
  - 여러마리의 강아지가 반환될 수 있도록 변경
- 강아지 진료기록 추가 함수 변경
  - mongodb알고리즘 추가
  - 이름/소유자이름/소유자번호를 patch url에서 받아서 정확하게 강아지를 찾을 수 있도록 함

qkrdmstlr3 committed:
[fix] API 호출경로 변경

- api호출을 dog에서 dogs로 경로 변경
- REST API에서 자원은 복수여야하기 때문이다

bardisue commented:
확인하였습니다. 변경사항 반영하여 진행하겠습니다.

2ternal commented:
확인했습니다. 수고하셨습니다.

qkrdmstlr3 committed:
[feat] mongodb사용을 위해 dependency추가

- mongodb사용을 위해서 dependency를 pom.xml파일에 추가
- parent set에 정의가 되어있어서 dependency를 적어주기만해도 상속할 수 있다

qkrdmstlr3 committed:
[refactor] DogRepository getDogByOwnerPhoneNumber함수 mongodb적용

- 주인의 전화번호로 강아지를 조회하는 getDogByOwnerPhoneNumber함수 변경
	- mongodb를 적용시켰다
- 테스트를 위해 insertDog를 간단하게 mongodb를 적용시켰다

qkrdmstlr3 committed:
[refactor] DogRepository modifyWithAddingDogRecord함수 mongodb적용

- 기존 강아지의 진료기록에 받은 기록을 추가하는 함수 수정
	- mongodb를 적용시켰다

qkrdmstlr3 committed:
[feat] 409에러 처리 위한 클래스 생성

- 409에러를 처리하기 위해서 클래스를 하나 생성함
- 강아지 생성 시 오류에 사용

qkrdmstlr3 committed:
[refactor] 강아지 생성/ 주인 전화번호로 조회 함수 수정

- 강아지 생성 시 이름/주인이름/주인전화번호가 unique하도록 검사
	- unique하다면 생성
	- 그렇지 않다면 409에러 반환

- 강아지 주인 전화번호로 조회 시 여러마리일 경우를 위해서 List를 반환하도록 변경

qkrdmstlr3 committed:
[refactor] 강아지의 진료기록을 추가하는 함수 변경

- unique key가 되는 이름/소유자이름/소유자전화번호 으로 강아지를 찾을 수 있도록 변경
	- 찾지 못했다면 404반환
- controller > service > repository로 넘기는 인자 수 증가

qkrdmstlr3 commented:
병준님이 말씀해주신 에러를 한 군데에서 관리하는 방법을 어떻게 해야될지 모르겠네요..

bardisue commented on src/main/java/org/cnu/realcoding/repository/DogRepository.java:
```diff
@@ -1,41 +1,53 @@
-    public void insertDog(Dog dog) {
-        dogs.add(dog);
+    public Boolean insertDog(Dog dog) {
+        Query query = new Query().addCriteria(
```
단일 개체를 생성하는 과정인데 addCriteria를 사용하신 이유가 있을까요?

YJR999 replied:
그러게요 addcriteria는 세부조건 추가할 때 쓰는거같은데요

qkrdmstlr3 replied:
@YJR999 
아 찾아보니 Query()안에 Criteria를 바로 넣는 방법도 있네요.

@bardisue 
이 query함수는 기존에 강아지가 있는지를 확인하기 위해서 조회에 사용됩니다. findAndModify는 있는데 findAndCreate함수가 없더라구요..

qkrdmstlr3 replied:
findAndCreate함수는 존재하지 않아서 조회 후에 삽입하는 방식으로 구현했습니다

qkrdmstlr3 committed:
[refactor] 강아지 생성 시 db에서 조회 함수 변경

- 강아지를 생성 시 db에서 조회할 때 Criteria객체를 Query객체 생성시 생성자에 전달해준다.
	- addCriteria는 몇가지 조건을 추가할 때 사용되는 것으로 보인다
